### PR TITLE
gh-126076: Account for relocated objects in tracemalloc

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-126076

### CODE REVIEW

**Comment by pablogsal** [comment]
CC @vstinner

File: Objects/tupleobject.c [hunk 1]
```diff
@@ -961,6 +961,7 @@ _PyTuple_Resize(PyObject **pv, Py_ssize_t newsize)
 #ifdef Py_TRACE_REFS
     _Py_ForgetReference((PyObject *) v);
 #endif
+    _PyReftracerTrack(v, PyRefTracer_DESTROY);
```
Comment (vstinner) [inline Objects/tupleobject.c]:
```suggestion
    _PyReftracerTrack((PyObject *)v, PyRefTracer_DESTROY);
```

Comment (pablogsal) [reply Objects/tupleobject.c]: Done

File: Include/internal/pycore_object.h [hunk 2]
```diff
@@ -94,6 +94,15 @@ PyAPI_FUNC(void) _Py_NO_RETURN _Py_FatalRefcountErrorFunc(
 #define _Py_FatalRefcountError(message) \
     _Py_FatalRefcountErrorFunc(__func__, (message))
 
+#define _PyReftracerTrack(obj, operation) \
+    do { \
+        struct _reftracer_runtime_state *tracer = &_PyRuntime.ref_tracer; \
+        if (tracer->tracer_func != NULL) { \
+            void* data = tracer->tracer_data; \
+            tracer->tracer_func(obj, (operation), data); \
```
Comment (vstinner) [inline Include/internal/pycore_object.h]:
```suggestion
            tracer->tracer_func((obj), (operation), data); \
```

**Review by vstinner — APPROVED** [review]
LGTM. But I suggest adding parenthesis in the macro implementation to avoid macro pitfals.

File: Include/internal/pycore_object.h [hunk 3]
```diff
@@ -94,6 +94,15 @@ PyAPI_FUNC(void) _Py_NO_RETURN _Py_FatalRefcountErrorFunc(
...
+        } \
+    } while(0)
+
```
Comment (vstinner) [inline Include/internal/pycore_object.h]: The linter wants to kill trailing spaces:

```suggestion
    } while(0)

```

**Comment by pablogsal** [comment]
> LGTM. But I suggest adding parenthesis in the macro implementation to avoid macro pitfals.

Will do

File: Objects/tupleobject.c [hunk 4]
```diff
@@ -961,6 +961,7 @@ _PyTuple_Resize(PyObject **pv, Py_ssize_t newsize)
 #ifdef Py_TRACE_REFS
     _Py_ForgetReference((PyObject *) v);
 #endif
+    _PyReftracerTrack((PyObject *)v, PyRefTracer_DESTROY);
```
Comment (markshannon) [inline Objects/tupleobject.c]: Can you move this down a few lines next to the call to `PyObject_GC_Resize`.
That way we can make sure it is eliminated when we fix https://github.com/python/cpython/issues/125790

File: Objects/object.c [hunk 5]
```diff
@@ -2917,12 +2910,7 @@ _Py_Dealloc(PyObject *op)
     // Make sure that type->tp_name remains valid
     Py_INCREF(type);
 #endif
-
-    struct _reftracer_runtime_state *tracer = &_PyRuntime.ref_tracer;
-    if (tracer->tracer_func != NULL) {
-        void* data = tracer->tracer_data;
-        tracer->tracer_func(op, PyRefTracer_DESTROY, data);
-    }
+    _PyReftracerTrack(op, PyRefTracer_DESTROY);
```
Comment (markshannon) [inline Objects/object.c]: Can you move this down a few lines next to the call to `(*dealloc)(op)`.
That way we can make sure it is eliminated when we fix https://github.com/python/cpython/issues/125790

File: Objects/object.c [hunk 6]
```diff
@@ -2525,10 +2521,7 @@ _Py_ResurrectReference(PyObject *op)
 #ifdef Py_TRACE_REFS
     _Py_AddToAllObjects(op);
 #endif
-    if (_PyRuntime.ref_tracer.tracer_func != NULL) {
-        void* data = _PyRuntime.ref_tracer.tracer_data;
-        _PyRuntime.ref_tracer.tracer_func(op, PyRefTracer_CREATE, data);
-    }
+    _PyReftracerTrack(op, PyRefTracer_CREATE);
```
Comment (markshannon) [inline Objects/object.c]: This seems incorrect.

The tracemalloc hook should trigger when the object is freed, not when its `tp_dealloc` is called.
So this object should still be live, IIRC.

**Review by markshannon — COMMENTED** [review]
Can you make sure that all calls to `_PyReftracerTrack` are directly before the relevant free, to make it easier to merge them when we fix https://github.com/python/cpython/issues/125790

**Comment by pablogsal** [comment]
> Can you make sure that all calls to `_PyReftracerTrack` are directly before the relevant free, to make it easier to merge them when we fix #125790

Yup!

File: Objects/object.c (hunk 5 above)
Comment (pablogsal) [reply Objects/object.c]: Sure, will do

File: Objects/tupleobject.c (hunk 4 above)
Comment (pablogsal) [reply Objects/tupleobject.c]: Sure, will do

File: Objects/object.c (hunk 6 above)
Comment (pablogsal) [reply Objects/object.c]: This is a creation call not a destruction call but I think you are right. There was no object really being "created" again, just a new reference.

Comment (pablogsal) [reply Objects/object.c]: Hummm but then we are we calling `_Py_AddToAllObjects` in this code path?

Comment (pablogsal) [reply Objects/object.c]: Ah, but indeed there is no "new object pointer" being created here. That's a good point.

File: Objects/tupleobject.c (hunk 4 above)
Comment (pablogsal) [reply Objects/tupleobject.c]: Done

File: Objects/object.c (hunk 5 above)
Comment (pablogsal) [reply Objects/object.c]: Done

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Relocated objects such as ``tuple``, ``bytes`` and ``unicode`` objects are
+properly tracked by :mod:`tracemalloc` and its associated hooks. Patch by
+Pablo Galindo
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst]:
```suggestion
Pablo Galindo.
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Relocated objects such as ``tuple``, ``bytes`` and ``unicode`` objects are
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst]: Should say "unicode" or "str" objects here? (from a end-user PoV, is there a difference here?)

File: Include/internal/pycore_object.h [hunk 9]
```diff
@@ -94,6 +94,14 @@ PyAPI_FUNC(void) _Py_NO_RETURN _Py_FatalRefcountErrorFunc(
 #define _Py_FatalRefcountError(message) \
     _Py_FatalRefcountErrorFunc(__func__, (message))
 
+#define _PyReftracerTrack(obj, operation) \
+    do { \
+        struct _reftracer_runtime_state *tracer = &_PyRuntime.ref_tracer; \
+        if (tracer->tracer_func != NULL) { \
+            void* data = tracer->tracer_data; \
```
Comment (picnixz) [inline Include/internal/pycore_object.h]:
```suggestion
            void *data = tracer->tracer_data; \
```

**Review by picnixz — COMMENTED** [review]
(I would have suggested using uppercase names for the macro but it wouldn't be consistent with the rest of the macros so I gave up)

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst (hunk 8 above)
Comment (pablogsal) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst]: We can say str objects yeah

**Comment by vstinner** [comment]
What's the status of this issue?

**Comment by pablogsal** [comment]
I just need to land it after checking that I have addressed everything (I was on vacation 😉)

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst (hunk 8 above)
Comment (pablogsal) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-28-13-18-16.gh-issue-126076.MebZuS.rst]:
```suggestion
Relocated objects such as ``tuple``, ``bytes`` and ``str`` objects are
```

**Pull request merged**